src/energy_meter.cpp

```diff
@@ -0,0 +1,508 @@
+        StaticJsonDocument<capacity> doc;
+        DeserializationError err = deserializeJson(doc, ret);
+        file.close();
+        DBUGVAR(err.code());
+        if (DeserializationError::Code::Ok == err)
+        {
+            _data.deserialize(doc);
+            DBUGVAR(_data.elapsed);
+            DBUGVAR(_data.total);
+            DBUGVAR(_data.session);
+            DBUGVAR(_data.daily);
+            DBUGVAR(_data.weekly);
+            DBUGVAR(_data.monthly);
+            DBUGVAR(_data.yearly);
+            DBUGVAR(_data.switches);
+            DBUGVAR(_data.date.day);
+            DBUGVAR(_data.date.month);
+            DBUGVAR(_data.date.year);
+            DBUGVAR(_data.imported);
+            // check if we need to reset some counters
+            rotate();
+            return true;
+        }
+        else
+        {
+            DBUGLN("EnergyMeter: Can't parse 'emeter.json', creating a new file");
+            _data.reset();
+            if (createEnergyMeterStorage(true, false))
+            {
+                return true;
+            }
+            else
+                return false;
+        }
+    }
+    else
+    {
+        file.close();
+        DBUGLN("Energy Meter: File missing, creating");
+        if (createEnergyMeterStorage(true, false))
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+};
+
+bool EnergyMeter::write(EnergyMeterData &data)
+{
+    DBUGLN("Energy Meter: Saving data");
+    File file = LittleFS.open(ENERGY_METER_FILE, "w", true);
+    if (!file)
+    {
+        file.close();
+        DBUGLN("Energy Meter: error can't open/create file");
+        return false;
+    }
+    StaticJsonDocument<capacity> doc;
+    _data.serialize(doc);
+    // Keep previous "imported" property
+    if (_data.imported)
+    {
+        doc["imported"] = _data.imported;
+    }
+    if (serializeJson(doc,file))
+    {
+        DBUGLN("Energy Meter: data saved");
+        file.close();
+        return true;
+    }
+    else
+    {
+        file.close();
+        DBUGLN("Energy Meter: can't write to file");
+        return false;
+    }
+};
+
+bool EnergyMeter::createEnergyMeterStorage(bool fullreset = false, bool forceimport = false)
+{
+    _data.reset(fullreset, forceimport);
+    _data.date = getCurrentDate();
+    if (forceimport) {
+        return _monitor->importTotalEnergy();
+    }
+    else {
+        return write(_data);
+    }
+};
+
+EnergyMeterDate EnergyMeter::getCurrentDate()
+{
+    EnergyMeterDate date;
+    struct timeval local_time;
+    gettimeofday(&local_time, NULL);
+    tm *timeinfo = gmtime(&local_time.tv_sec);
+    uint16_t yday = (uint16_t)timeinfo->tm_yday;
+    date.day = (uint8_t)timeinfo->tm_wday;
+    date.month = (uint8_t)timeinfo->tm_mon;
+    date.year = 1900 + (uint16_t)timeinfo->tm_year;
+
+    return date;
+};
+
+void EnergyMeter::increment_switch_counter()
+{
+  if (_switch_state != _monitor->isActive())
```

Comment:
Just spotted a bug, I think this should be isCharging to reflect switches of the relay, isActive will increent the count in other cases, eg if the car is not plugged it and the state toggeles from active to disabled.